api/handlers.go

```diff
@@ -1225,6 +1235,25 @@ func (api *API) checkUserQuotas(ctx context.Context, u *database.User) {
 	}
 }
 
+// userFromRequest checks the requests for various forms of authentication (API
+// key, cookie, authorization header) and returns user information based on
+// those.
+func (api *API) userFromRequest(req *http.Request) (*database.User, jwt2.Token, error) {
+	// Check for an API key.
+	u, tk, err := api.userAndTokenByAPIKey(req)
+	if err != nil && !errors.Contains(err, ErrNoAPIKey) {
+		return nil, nil, err
+	}
+	// If there is no API key check for a token.
+	if errors.Contains(err, ErrNoAPIKey) {
+		u, tk, err = api.userAndTokenByRequestToken(req)
+		if err != nil {
+			return nil, nil, err
+		}
```

Comment:
Suggestion:

```go
if errors.Contains(err, ErrNoAPIKey) {
  return api.userAndTokenByRequestToken(req)
}
return u, tk, err
```

Reply:
I agree with this approach. I've gotten some flak in reviews for using it, though.
Will add it in a F/U.